### PR TITLE
feat: add isqrt built-in

### PR DESCRIPTION
**Pull request by bout3fiddy**
### What I did

Inspired from [t11s's solmate fixed point sqrt utility algo].(https://github.com/transmissions11/solmate/blob/9cf1428245074e39090dceacb0c28b1f684f584c/src/utils/FixedPointMathLib.sol#L166)

Michael from Curve made a version of this in vyper. I added it as a builtin and changed a few tests.

### How I did it

Take code from Michael, amend a few things, write tests

### How to verify it

`pytest tests/parser/types/numbers/test_sqrt_solmate.py`

### Commit message

Adds `isqrt` builtin. For a better explanation of the algorithm, refer to: [https://github.com/Gaussian-Process/solmate/blob/837de01395312eb89e607fdc64fb7bb9c03207c3/src/utils/FixedPointMathLib.sol](https://github.com/Gaussian-Process/solmate/blob/837de01395312eb89e607fdc64fb7bb9c03207c3/src/utils/FixedPointMathLib.sol)

### Cute Animal Picture

rottweiler puppy

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/builtin_functions/functions.py [hunk 1]
```diff
@@ -2621,6 +2689,7 @@ def evaluate(self, node):
     "pow_mod256": PowMod256(),
     "uint2str": Uint2Str(),
     "sqrt": Sqrt(),
+    "sqrt_solmate": SqrtSolmate(),
```
Comment (fubuloubu) [inline vyper/builtin_functions/functions.py:2692]: why not update the original Sqrt instead?

Comment (bout3fiddy) [reply vyper/builtin_functions/functions.py:2692]: That was the original idea indeed, but did not want to introduce that breaking change without prompt from a core contributor.

I'll make the changes to the original sqrt then.

Comment (charles-cooper) [reply vyper/builtin_functions/functions.py:2692]: i don't think we should get rid of the original sqrt. sqrt on decimals is pretty useful, especially once we open up the decimals type to be generic.

Comment (fubuloubu) [reply vyper/builtin_functions/functions.py:2692]: Yeah, but sounds like this should basically expand the types that sqrt works with, so if it's decimals it uses the previous and if it's integers it uses this one

Comment (charles-cooper) [reply vyper/builtin_functions/functions.py:2692]: i think if it's possible, use the same algorithm for both types. there will be fewer gas surprises that way

Comment (bout3fiddy) [reply vyper/builtin_functions/functions.py:2692]: @charles-cooper It doesn't look possible since exponentiation is not allowed for decimal type. I do like the idea @fubuloubu mentioned, where `sqrt(x: decimal)` would just yield what it does right now and `sqrt(x: uint256)` would use the optimised version. 

Regarding gas surprises: is that really an issue? I don't expect users to use the decimal type at all.

**Comment by charles-cooper** [comment]
looks like `isqrt` is not available in python 3.7. we probably need to backport, @bout3fiddy see https://github.com/vyperlang/vyper/blob/e7733e15003816d7a323476678a2ac7b3755628f/vyper/utils.py#L40-L44 for the pattern

**Review by charles-cooper — APPROVED** [review]
very nice pull request. i think the only thing i would change is in the test file, changing `from math import isqrt` to `import math`. that way the usages must be qualified as `math.isqrt` which helps the reader distinguish from vyper isqrt.

File: docs/built-in-functions.rst [hunk 2]
```diff
@@ -664,6 +664,22 @@ Math
         >>> ExampleContract.foo(9.0)
         3.0
 
+.. py:function:: isqrt(a: uint256) -> uint256
+
+    Return the square root of the provided decimal number, using `Solmate's square root algorithm <https://github.com/transmissions11/solmate/blob/9cf1428245074e39090dceacb0c28b1f684f584c/src/utils/FixedPointMathLib.sol#L166>`_.
```
Comment (charles-cooper) [inline docs/built-in-functions.rst:669]: probably the best place for this attribution is in the comments or commit message. having it in the documentation might lead to an expectation that the semantics or performance are identical to solmate, which we don't want to imply. in the documentation itself we might want to say what the general method is (babylonian right?) and any semantics that are important (i think we do want to mention that it rounds down)

File: docs/built-in-functions.rst [hunk 3]
```diff
@@ -664,6 +664,22 @@ Math
...
+    .. code-block:: python
+
+        @external
+        @view
+        def foo(a: uint256) -> uint256:
+            return isqrt(a)
+
+    .. code-block:: python
+
+        >>> ExampleContract.foo(9)
```
Comment (charles-cooper) [inline docs/built-in-functions.rst:680]: an example of round down behavior would be good here

File: vyper/builtin_functions/functions.py [hunk 4]
```diff
@@ -2190,6 +2191,74 @@ def build_IR(self, expr, args, kwargs, context):
         )
 
 
+class IntegerSqrt(BuiltinFunction):
```
Comment (charles-cooper) [inline vyper/builtin_functions/functions.py:2194]: maybe it's clearer to rename this class to `ISqrt`

File: docs/built-in-functions.rst [hunk 5]
```diff
@@ -664,6 +664,25 @@ Math
         >>> ExampleContract.foo(9.0)
         3.0
 
+.. py:function:: isqrt(x: uint256) -> uint256
+
+    Return the square root of the provided integer number, using Babylonian square root algorithm.
```
Comment (charles-cooper) [inline docs/built-in-functions.rst:669]:
```suggestion
    Return the (integer) square root of the provided integer number, using the Babylonian square root algorithm. The rounding mode is to round down to the nearest integer. For instance, ``isqrt(101) == 10``.
```

File: docs/built-in-functions.rst [hunk 6]
```diff
@@ -664,6 +664,25 @@ Math
...
+
+    .. note::
+        Performance note: If ``x+1`` is a perfect square, then the Babylonian algorithm oscillates between floor(sqrt(x)) and ceil(sqrt(x)) in consecutive iterations. ``isqrt`` has a final check that returns the floor value always, but this increases costs by approximately 10%.
```
Comment (charles-cooper) [inline docs/built-in-functions.rst:672]: can the user configure this? adding this as a performance note makes it sounds like they can -- this comment seems more directed at future maintainers of the implementation, in which case it would make more sense as comments in the implementation source code.

Comment (bout3fiddy) [reply docs/built-in-functions.rst:672]: a user can configure this indeed. we could set up a `precise = True` flag in the `isqrt` builtin if you think that helps users shave off some gas?

Comment (charles-cooper) [reply docs/built-in-functions.rst:672]: I think being configurable is not worth it at this point and we should move the note to the source code (for future maintainers).

File: tests/parser/types/numbers/test_isqrt.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+@pytest.mark.fuzzing
+@hypothesis.given(
+    value=hypothesis.strategies.integers(min_value=0, max_value=SizeLimits.MAX_UINT256)
+)
+@hypothesis.example(SizeLimits.MAX_UINT256)
+@hypothesis.example(0)
+@hypothesis.example(1)
+@hypothesis.example(2704)
+@hypothesis.example(110889)
+@hypothesis.example(32239684)
```
Comment (charles-cooper) [inline tests/parser/types/numbers/test_isqrt.py:119]: maybe add a note that these values are perfect squares + 1

Comment (charles-cooper) [reply tests/parser/types/numbers/test_isqrt.py:119]:
```suggestion
# these examples demonstrate correct rounding mode
# for an edge case in the babylonian method - the operand is
# a perfect square + 1
@hypothesis.example(2704)
@hypothesis.example(110889)
@hypothesis.example(32239684)
```

File: vyper/builtin_functions/functions.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/builtin_functions/functions.py]: discussed offline -- the gas difference is ok. @bout3fiddy added a note that we will try to reify the code paths later

File: tests/parser/types/numbers/test_isqrt.py (hunk 7 above)
Comment (charles-cooper) [reply tests/parser/types/numbers/test_isqrt.py:119]: correction: perfect squares - 1

**Comment by charles-cooper** [comment]
> looks like `isqrt` is not available in python 3.7. we probably need to backport

no more need to backport, 3.7 tests were dropped in https://github.com/vyperlang/vyper/pull/3071

**Review by charles-cooper — APPROVED** [review]
